core/script.runtime/src/kotlin/script/dependencies/resolvers.kt

```diff
@@ -32,6 +50,7 @@ interface DependenciesResolver : @Suppress("DEPRECATION") ScriptDependenciesReso
 
     sealed class ResolveResult {
```

Comment:
My main concern with your proposal is that the user has to pass both `isSuccess` flag and notnull `dependencies` in both cases. While `ScriptDependencies.Empty` is a completely valid value for a successful resolution, it's not clear that it's meant to be ignored if `isSuccessful == false`. That could be explained in the docs of course.
I agree that using sealed class makes API a bit cumbursome. What I find problematic in the current API is that it's not clear that successful resolution can actually contain error diagnostics and that may be confusing for implementors of the API. 
I would not put a lot of emphasis on ease of consuming the API (since it's gonna be us who'll be doing that) but rather focus on the API being easy to implement.
In my mind those snippets would look like (with the current API):
```
when (result) {
    is Success -> ...
    is Failure -> ...
}
```
or 
```
val dependencies = result.dependencies ?: ScriptDependencies.Empty
val diagnostics = result.diagnostics
...
```